File: stacks/IStackDemo.java
package stacks;

import stacks.impl.StackUsingArrayList;
import stacks.impl.StackUsingLinkedList;

public class IStackDemo {
    public static void main(String[] args) {
        System.out.println("======= Stack using ArrayList =======");
        baseDemo(new StackUsingArrayList<>());
        System.out.println();
        System.out.println();

        System.out.println("======= Stack using LinkedList =======");
        baseDemo(new StackUsingLinkedList<>());
        System.out.println();
        System.out.println();
    }

    static void baseDemo(IStack<Integer> iStack) {
        int end = 5;
        System.out.println("Pushing 1 till " + end + " to stack");
        for (int i = 1; i <= end; i++) {
            iStack.push(i);
            System.out.println("Stack: " + iStack);
        }

        System.out.println();
        System.out.println("Popping till stack is empty");
        while (!iStack.isEmpty()) {
            int item = iStack.peek();
            iStack.pop();
            System.out.println("Popped: " + item);
            System.out.println("Stack: " + iStack);
        }
    }
}

File: src/stacks/IStackDemo.kt
package src.stacks

import src.stacks.impl.StackUsingArrayList
import src.stacks.impl.StackUsingLinkedList

object IStackDemo {

    @JvmStatic
    fun main(args: Array<String>) {
        println("======= Stack using ArrayList =======")
        baseDemo(StackUsingArrayList())
        println()

        println()
        println("======= Stack using LinkedList =======")
        baseDemo(StackUsingLinkedList())
        println()
        println()
    }

    private fun baseDemo(iStack: IStack<Int?>) {
        val end = 5
        println("Pushing 1 till $end to stack")
        for (i in 1..end) {
            iStack.push(i)
            println("Stack: $iStack")
        }

        println()
        println("Popping till stack is empty")
        while (!iStack.isEmpty()) {
            val item = iStack.peek()!!
            iStack.pop()
            println("Popped: $item")
            println("Stack: $iStack")
        }
    }
}
